base/float.jl

```diff
@@ -414,9 +422,122 @@ muladd(x::T, y::T, z::T) where {T<:IEEEFloat} = muladd_float(x, y, z)
+    number <<= lz
+    epint -= lz
+    bits = T(0)
+    if epint >= T(0)
+        mant = number & significand_mask(Tfloat) #this might not be necessary
+        bits |= mant
+        epint = ((epint + Tint(1)) << mantissa_width(Tfloat)) & exponent_mask(Tfloat)
+        bits |= epint
+    else
+        number = number >> -epint
+        mant = number & significand_mask(Tfloat)
+        bits |= mant
+    end
+    return reinterpret(Tfloat, bits)
+end
+
+function fmod_internal(x::T, y::T) where {T<:IEEEFloat}
+    xuint = reinterpret(Unsigned, x)
+    yuint = reinterpret(Unsigned, y)
+    if xuint <= yuint
+        if xuint < yuint
+            return x
+        end
+        return zero(T)
+    end
+
+    e_x = unbiased_exponent(x)
+    e_y = unbiased_exponent(y)
+    # Most common case where |y| is "very normal" and |x/y| < 2^EXPONENT_WIDTH
+    if e_y > mantissa_width(T) && (e_x - e_y) <= exponent_width(T)
+        m_x = explicit_mantissa_noinfnan(x)
+        m_y = explicit_mantissa_noinfnan(y)
+        d = (e_x == e_y) ? (m_x - m_y) : (m_x << (e_x - e_y)) % m_y
+        iszero(d) && return zero(T)
+        return make_value(d, e_y - uinttype(T)(1))
+    end
+    # Both are subnormals
+    if e_x == uinttype(T)(0) && e_y == uinttype(T)(0)
+        bits = uinttype(T)(0)
+        number = xuint % yuint
+        mant = number & significand_mask(T) #this might not be necessary
+        bits |= mant
+        return reinterpret(T, bits)
+    end
+
+    m_x = explicit_mantissa_noinfnan(x)
+    e_x -= uinttype(T)(1)
+    m_y = explicit_mantissa_noinfnan(y)
+    lz_m_y = exponent_width(T)
+    if e_y > uinttype(T)(0)
+        e_y -= uinttype(T)(1)
+    else
+        m_y = mantissa(y)
+        lz_m_y = Core.Intrinsics.ctlz_int(m_y)
+    end
+
+    tz_m_y = Core.Intrinsics.cttz_int(m_y)
+    sides_zeroes_cnt = lz_m_y + tz_m_y
+
+    # n>0
+    exp_diff::uinttype(T) = e_x - e_y
+    # Shift hy right until the end or n = 0
+    right_shift = exp_diff < tz_m_y ? exp_diff : tz_m_y
+    m_y >>= right_shift
+    exp_diff -= right_shift
+    e_y += right_shift
+    # Shift hx left until the end or n = 0
+    left_shift = exp_diff < exponent_width(T) ? exp_diff : exponent_width(T)
+    m_x <<= left_shift
+    exp_diff -= left_shift
+
+    m_x %= m_y
+    iszero(m_x) && return zero(T)
+    iszero(exp_diff) && return make_value(m_x, e_y)
+
+    while exp_diff > sides_zeroes_cnt
+        exp_diff -= sides_zeroes_cnt
+        m_x <<= sides_zeroes_cnt
+        m_x %= m_y
+    end
+    m_x <<= exp_diff
+    m_x %= m_y
+    return make_value(m_x, e_y)
+end
+
+function fmod(x::T, y::T) where {T<:IEEEFloat}
+    if (!isnan(x) && isfinite(x) && !iszero(x) && !isnan(y) && isfinite(y) && !iszero(y))
+        xabs = abs(x)
+        yabs = abs(y)
+        res = fmod_internal(xabs, yabs)
+        return copysign(res,x)
+    end
+    (isnan(x) || isnan(y)) && return T(NaN)
+    (isinf(x) || iszero(y)) && return T(NaN)
+    isinf(y) && return x
+    return x
```

Comment:
```suggestion
    return x
```
We're returning `x` either way